File: api/schemas/batch_validation.py
"""Schémas Pydantic pour la validation batch FR87."""

from __future__ import annotations

from typing import List, Literal, Optional

from pydantic import BaseModel, Field


class BatchValidationIssueResponse(BaseModel):
    """Issue de validation agrégée."""

    type: str
    message: str
    severity: Literal["error", "warning"]
    source: Literal["graph", "schema", "lore"]
    node_id: Optional[str] = None


class BatchValidationItemResponse(BaseModel):
    """Résultat pour un dialogue."""

    document_id: str
    status: Literal["valid", "invalid", "skipped", "cancelled", "denied"]
    issues: List[BatchValidationIssueResponse] = Field(default_factory=list)
    validated_at: str = ""


class BatchValidateRequest(BaseModel):
    """Requête sync (N < 20)."""

    document_ids: List[str] = Field(..., min_length=1, max_length=19)


class BatchValidateResponse(BaseModel):
    """Rapport sync complet."""

    items: List[BatchValidationItemResponse]
    cancelled: bool = False
    started_at: str
    finished_at: str
    valid_count: int
    invalid_count: int
    skipped_count: int


class BatchValidateJobRequest(BaseModel):
    """Requête job async (N ≥ 20)."""

    document_ids: List[str] = Field(..., min_length=20, max_length=500)


class BatchValidateJobCreateResponse(BaseModel):
    """Création de job."""

    job_id: str
    status: str
    total: int


class BatchValidateJobStatusResponse(BaseModel):
    """État d'un job de validation batch."""

    job_id: str
    status: Literal["queued", "running", "completed", "cancelled", "error"]
    current: int = 0
    total: int = 0
    cancelled: bool = False
    error: Optional[str] = None
    report: Optional[BatchValidateResponse] = None

File: frontend/src/api/batchValidation.ts
/** Client API validation batch FR87. */
import apiClient from './client'

export type BatchValidationIssue = {
  type: string
  message: string
  severity: 'error' | 'warning'
  source: 'graph' | 'schema' | 'lore'
  node_id?: string | null
}

export type BatchValidationItem = {
  document_id: string
  status: 'valid' | 'invalid' | 'skipped' | 'cancelled' | 'denied'
  issues: BatchValidationIssue[]
  validated_at: string
}

export type BatchValidateReport = {
  items: BatchValidationItem[]
  cancelled: boolean
  started_at: string
  finished_at: string
  valid_count: number
  invalid_count: number
  skipped_count: number
}

export type BatchValidateJobStatus = {
  job_id: string
  status: 'queued' | 'running' | 'completed' | 'cancelled' | 'error'
  current: number
  total: number
  cancelled: boolean
  error?: string | null
  report?: BatchValidateReport | null
}

export const BATCH_VALIDATE_JOB_MIN = 20

export async function batchValidateSync(
  documentIds: string[],
  signal?: AbortSignal
): Promise<BatchValidateReport> {
  const response = await apiClient.post<BatchValidateReport>(
    '/api/v1/dialogues/batch-validate',
    { document_ids: documentIds },
    { signal }
  )
  return response.data
}

export async function startBatchValidateJob(
  documentIds: string[]
): Promise<{ job_id: string; status: string; total: number }> {
  const response = await apiClient.post<{
    job_id: string
    status: string
    total: number
  }>('/api/v1/dialogues/batch-validate/jobs', { document_ids: documentIds })
  return response.data
}

export async function getBatchValidateJob(
  jobId: string
): Promise<BatchValidateJobStatus> {
  const response = await apiClient.get<BatchValidateJobStatus>(
    `/api/v1/dialogues/batch-validate/jobs/${jobId}`
  )
  return response.data
}

export async function cancelBatchValidateJob(
  jobId: string
): Promise<BatchValidateJobStatus> {
  const response = await apiClient.post<BatchValidateJobStatus>(
    `/api/v1/dialogues/batch-validate/jobs/${jobId}/cancel`
  )
  return response.data
}
